### PR TITLE
Feature/tao 4118 normal maximum calculation

### PR DESCRIPTION
**Pull request by ssipasseuth**
Added MAXSCORE calculation in item creator + fixed some interaction in authoring

### CODE REVIEW

File: views/js/qtiItem/core/Item.js [hunk 1]
```diff
@@ -83,6 +83,19 @@ define([
             }
             return this;
         },
+        getOutcomeDeclaration : function getOutcomeDeclaration(identifier){
+            var found;
+            _.each(this.outcomes, function (outcome) {
```
Comment (jsconan) [inline views/js/qtiItem/core/Item.js]: `_.each()` is deprecated, please use `_.forEach()` instead.
You could also use `_.find()`

**Comment by ssipasseuth** [comment]
fixed various issues, please check again

**Comment by ssipasseuth** [comment]
@jsconan the feature has been approved by PM and QA. You can proceed to the last code review now.
Thanks !

File: views/js/qtiItem/helper/maxScore.js [hunk 2]
```diff
@@ -0,0 +1,738 @@
...
+            return max;
+        },
+
+        /**
+         * Compute the maximum score of a "order" typed interaction
+         * @param {Object} interaction - a standard interaction model object
+         * @returns {Number}
+         */
+        orderInteractionBased : function orderInteractionBased(interaction){
+            var minChoice = _ignoreMinChoice ? 0 : parseInt(interaction.attr('minChoices')||0);
```
Comment (jsconan) [inline views/js/qtiItem/helper/maxScore.js]: `parseInt` without `base` parameter

File: views/js/qtiItem/helper/maxScore.js [hunk 3]
```diff
@@ -0,0 +1,738 @@
...
+            var maxChoice = parseInt(interaction.attr('maxChoices')||0);
```
Comment (jsconan) [inline views/js/qtiItem/helper/maxScore.js]: same here

File: views/js/qtiItem/helper/maxScore.js [hunk 4]
```diff
@@ -0,0 +1,738 @@
...
+
+        /**
+         * Compute the maximum score of a "associate" typed interaction
+         * @param {Object} interaction - a standard interaction model object
+         * @returns {Number}
+         */
+        associateInteractionBased : function associateInteractionBased(interaction, options){
+            var responseDeclaration = interaction.getResponseDeclaration();
+            var template = responseHelper.getTemplateNameFromUri(responseDeclaration.template);
+            var maxAssoc = parseInt(interaction.attr('maxAssociations')||0);
```
Comment (jsconan) [inline views/js/qtiItem/helper/maxScore.js]: missing `base` parameter

File: views/js/qtiItem/helper/maxScore.js [hunk 5]
```diff
@@ -0,0 +1,738 @@
...
+
+                    //check if the choices usage are possible within the constraint defined in the interaction
+                    _.forEach(_.countBy(choicesIdentifiers), function(count, identifier){
+                        var matchMax;
+                        var choice = interaction.getChoiceByIdentifier(identifier);
+                        if(!choice){
+                            max = 0;
+                            return false;
+                        }
+                        matchMax = parseInt(choice.attr('matchMax'));
```
Comment (jsconan) [inline views/js/qtiItem/helper/maxScore.js]: missig `base` parameter

File: views/js/qtiItem/helper/maxScore.js [hunk 6]
```diff
@@ -0,0 +1,738 @@
...
+                        choices = pair.trim().split(' ');
+                        if(_.isArray(choices) && choices.length === 2){
+                            group1.push(choices[0].trim());
+                            group2.push(choices[1].trim());
+                        }
+                    });
+
+                    _.forEach(_.countBy(group1), function(count, identifier){
+                        var choice = interaction.getChoiceByIdentifier(identifier);
+                        var matchMax = parseInt(choice.attr('matchMax'));
```
Comment (jsconan) [inline views/js/qtiItem/helper/maxScore.js]: missing `base` parameter

File: views/js/qtiItem/helper/maxScore.js [hunk 7]
```diff
@@ -0,0 +1,738 @@
...
+                        gapId = choices[1];
+                        if(!_usedChoices[choiceId]){
+                            choice = interaction.getChoiceByIdentifier(choiceId);
+                            if(!choice){
+                                //inexisting choice, skip
+                                return false;
+                            }
+                            _usedChoices[choiceId] = {
+                                used : 0,
+                                max: parseInt(choice.attr('matchMax'))
```
Comment (jsconan) [inline views/js/qtiItem/helper/maxScore.js]: missing `base` parameter

File: views/js/qtiItem/helper/maxScore.js [hunk 8]
```diff
@@ -0,0 +1,738 @@
...
+
+                //sort the score map entries by the score
+                scoreMaps = _.values(responseDeclaration.mapEntries);
+                sortedMapEntries = _(scoreMaps).map(function (v) {
+                    return parseFloat(v);
+                }).sortBy().reverse().first(totalAnswerableResponse);
+
+                //if there is not enough map defined, compared to the minChoice constraint, fill in the rest of required choices with the default map
+                missingMapsCount = minChoice - sortedMapEntries.size();
+                for(i = 0; i < missingMapsCount;i++){
```
Comment (jsconan) [inline views/js/qtiItem/helper/maxScore.js]: you could also use the lodash helper: `_.times()`

File: views/js/qtiItem/helper/maxScore.js [hunk 9]
```diff
@@ -0,0 +1,738 @@
...
+
+                //if the map default is positive, the optimal strategy involves using as much mapDefault as possible
+                if(mapDefault && mapDefault > 0){
+                    if(maxChoice){
+                        missingMapsCount = maxChoice - sortedMapEntries.size();
+                    }else{
+                        missingMapsCount = _.size(interaction.getChoices()) - sortedMapEntries.size();
+                    }
+                    if(missingMapsCount > 0){
+                        for(i = 0; i < missingMapsCount;i++){
```
Comment (jsconan) [inline views/js/qtiItem/helper/maxScore.js]: you could also use the lodash helper: `_.times()`

**Pull request merged**